File: src/Algorithms/Lists/Two_Numbers_Sum/two_numbers_sum.py


# O(n^2) time
# O(1) space
def find_sum(array, targetSum):
    for i in range(len(array) - 1):
        firstNum = array[i]
        for j in range(i+1, len(array)):
            secondNum = array[j]
            if firstNum + secondNum == targetSum:
                return [firstNum, secondNum]
    return -1


# O(nlog(n)) time (depends on a sorting algorithm)
# O(n) space (because we clone sorted array, otherwise it's O(1))
def sort_and_find_sum(array, targetSum):
    sortedArray = array.copy()
    sortedArray.sort()
    pointerA = 0
    pointerB = len(sortedArray) - 1

    while pointerA < pointerB:
        currentSum = sortedArray[pointerA] + sortedArray[pointerB]
        if currentSum == targetSum:
            return [sortedArray[pointerA], sortedArray[pointerB]]
        elif currentSum > targetSum:
            pointerB -= 1
        else:
            pointerA += 1

    return -1


# O(n) time
# O(n) space
def find_sum_with_set(array, targetSum):
    nums = {}
    for num in array:
        if targetSum - num in nums:
            return [targetSum - num, num]
        else:
            nums[num] = True
    return -1

File: src/Algorithms/Lists/Two_Numbers_Sum/two-numbers-sum.ts
// O(n^2) time
// O(1) space
const findSum = (array: number[], sum: number) => {
  let pointerA = 0;
  let pointerB = 1;

  while (pointerA < array.length - 1) {
    if (array[pointerA] + array[pointerB] === sum) {
      return [array[pointerA], array[pointerB]];
    }

    if (pointerB === array.length - 1) {
      pointerA++;
      pointerB = pointerA + 1;
    } else {
      pointerB++;
    }
  }

  return -1;
};

// Time O(n*log(n)) (depends on a sorting algorithm)
// Space O(n) (because we clone sorted array, otherwise it's O(1))
const sortAndFindSum = (array: number[], sum: number) => {
  const sortedArray = [...array].sort((a, b) => a - b);
  let pointerA = 0;
  let pointerB = array.length - 1;

  while (pointerA < pointerB) {
    const currentSum = sortedArray[pointerA] + sortedArray[pointerB];
    if (currentSum === sum) {
      return [sortedArray[pointerA], sortedArray[pointerB]];
    }

    if (currentSum > sum) {
      pointerB--;
    } else {
      pointerA++;
    }
  }

  return -1;
};

// Time O(n)
// Space O(n)
const findSumWithSet = (array: number[], sum: number) => {
  const set = new Set();
  let result: number | [number, number] = -1;

  array.forEach((value) => {
    const diff = sum - value;
    if (set.has(diff)) {
      result = [diff, value];
    }

    set.add(value);
  });

  return result;
};

export const solutions = {
  findSum,
  sortAndFindSum,
  findSumWithSet,
};
